File: backend/internal/models/flashcard.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// Deck represents a flashcard deck
type Deck struct {
	ID            uuid.UUID  `json:"id" db:"id"`
	UserID        uuid.UUID  `json:"user_id" db:"user_id"`
	Name          string     `json:"name" db:"name"`
	Description   *string    `json:"description" db:"description"`
	Topic         *string    `json:"topic" db:"topic"`
	Difficulty    *string    `json:"difficulty" db:"difficulty"`
	Color         *string    `json:"color" db:"color"`
	TotalCards    int        `json:"total_cards" db:"total_cards"`
	MasteredCards int        `json:"mastered_cards" db:"mastered_cards"`
	AverageScore  float64    `json:"average_score" db:"average_score"`
	StudyTime     int        `json:"study_time" db:"study_time"` // in minutes
	LastStudied   *time.Time `json:"last_studied" db:"last_studied"`
	CreatedAt     time.Time  `json:"created_at" db:"created_at"`
	UpdatedAt     time.Time  `json:"updated_at" db:"updated_at"`
}

// Flashcard represents a single flashcard
type Flashcard struct {
	ID            uuid.UUID  `json:"id" db:"id"`
	DeckID        uuid.UUID  `json:"deck_id" db:"deck_id"`
	Front         string     `json:"front" db:"front"`
	Back          string     `json:"back" db:"back"`
	Difficulty    string     `json:"difficulty" db:"difficulty"`
	TimesReviewed int        `json:"times_reviewed" db:"times_reviewed"`
	TimesCorrect  int        `json:"times_correct" db:"times_correct"`
	Mastery       string     `json:"mastery" db:"mastery"`
	LastReviewed  *time.Time `json:"last_reviewed" db:"last_reviewed"`
	NextReview    *time.Time `json:"next_review" db:"next_review"`
	Tags          []string   `json:"tags" db:"tags"`
	CreatedAt     time.Time  `json:"created_at" db:"created_at"`
	UpdatedAt     time.Time  `json:"updated_at" db:"updated_at"`
}

// StudySession represents a study session
type StudySession struct {
	ID             uuid.UUID  `json:"id" db:"id"`
	UserID         uuid.UUID  `json:"user_id" db:"user_id"`
	DeckID         uuid.UUID  `json:"deck_id" db:"deck_id"`
	Mode           string     `json:"mode" db:"mode"`
	CardsStudied   int        `json:"cards_studied" db:"cards_studied"`
	CorrectAnswers int        `json:"correct_answers" db:"correct_answers"`
	TotalTime      int        `json:"total_time" db:"total_time"` // in seconds
	StartedAt      time.Time  `json:"started_at" db:"started_at"`
	CompletedAt    *time.Time `json:"completed_at" db:"completed_at"`
	CreatedAt      time.Time  `json:"created_at" db:"created_at"`
}

// CreateDeckRequest represents the request to create a new deck
type CreateDeckRequest struct {
	Name        string  `json:"name" binding:"required,min=1,max=255"`
	Description *string `json:"description"`
	Topic       *string `json:"topic"`
	Difficulty  *string `json:"difficulty"`
	Color       *string `json:"color"`
}

// UpdateDeckRequest represents the request to update a deck
type UpdateDeckRequest struct {
	Name        *string `json:"name"`
	Description *string `json:"description"`
	Topic       *string `json:"topic"`
	Difficulty  *string `json:"difficulty"`
	Color       *string `json:"color"`
}

// CreateFlashcardRequest represents the request to create a new flashcard
type CreateFlashcardRequest struct {
	Front      string    `json:"front" binding:"required,min=1"`
	Back       string    `json:"back" binding:"required,min=1"`
	Difficulty *string   `json:"difficulty"`
	Tags       *[]string `json:"tags"`
}

// UpdateFlashcardRequest represents the request to update a flashcard
type UpdateFlashcardRequest struct {
	Front      *string   `json:"front"`
	Back       *string   `json:"back"`
	Difficulty *string   `json:"difficulty"`
	Tags       *[]string `json:"tags"`
}

// CreateBulkFlashcardsRequest represents the request to create multiple flashcards
type CreateBulkFlashcardsRequest struct {
	Cards []CreateFlashcardRequest `json:"cards" binding:"required,min=1,dive"`
}

// StudyCardRequest represents a card study action
type StudyCardRequest struct {
	CardID    uuid.UUID `json:"card_id" binding:"required"`
	IsCorrect bool      `json:"is_correct"`
	TimeSpent int       `json:"time_spent"` // in seconds
}

// StartStudySessionRequest represents the request to start a study session
type StartStudySessionRequest struct {
	DeckID uuid.UUID `json:"deck_id" binding:"required"`
	Mode   *string   `json:"mode"`
}

// EndStudySessionRequest represents the request to end a study session
type EndStudySessionRequest struct {
	CardsStudied   int                `json:"cards_studied"`
	CorrectAnswers int                `json:"correct_answers"`
	TotalTime      int                `json:"total_time"`
	StudiedCards   []StudyCardRequest `json:"studied_cards"`
}

// DeckWithStats represents a deck with additional statistics
type DeckWithStats struct {
	Deck
	Outstanding int `json:"outstanding"` // Cards due for review
	New         int `json:"new"`         // New cards not yet studied
}

// FlashcardStats represents statistics for flashcards
type FlashcardStats struct {
	TotalDecks       int     `json:"total_decks"`
	TotalCards       int     `json:"total_cards"`
	MasteredCards    int     `json:"mastered_cards"`
	CardsToReview    int     `json:"cards_to_review"`
	AverageScore     float64 `json:"average_score"`
	TotalStudyTime   int     `json:"total_study_time"`   // in minutes
	StudyStreak      int     `json:"study_streak"`       // days
	LastStudySession *time.Time `json:"last_study_session"`
}

// NewEmptyFlashcardStats returns a FlashcardStats with zero values
func NewEmptyFlashcardStats() *FlashcardStats {
	return &FlashcardStats{
		TotalDecks:       0,
		TotalCards:       0,
		MasteredCards:    0,
		CardsToReview:    0,
		AverageScore:     0,
		TotalStudyTime:   0,
		StudyStreak:      0,
		LastStudySession: nil,
	}
}

File: frontend/src/services/flashcard.ts
import { apiService } from "./api";

// Types
export interface Deck {
  id: string;
  user_id: string;
  name: string;
  description?: string;
  topic?: string;
  difficulty?: "easy" | "medium" | "hard";
  color?: string;
  total_cards: number;
  mastered_cards: number;
  average_score: number;
  study_time: number; // in minutes
  last_studied?: string;
  created_at: string;
  updated_at: string;
  // Optional UI properties
  outstanding?: number;
  new?: number;
}

export interface Flashcard {
  id: string;
  deck_id: string;
  front: string;
  back: string;
  difficulty?: "easy" | "medium" | "hard";
  mastery_level: number;
  times_reviewed: number;
  last_reviewed?: string;
  next_review?: string;
  created_at: string;
  updated_at: string;
}

export interface StudySession {
  id: string;
  deck_id: string;
  user_id: string;
  start_time: string;
  end_time?: string;
  total_cards: number;
  correct_answers: number;
  total_time?: number; // in seconds
  created_at: string;
}

export interface FlashcardStats {
  total_decks: number;
  total_cards: number;
  mastered_cards: number;
  cards_due_today: number;
  average_score: number;
  total_study_time: number; // in minutes
  longest_streak: number;
  current_streak: number;
}

// Request types
export interface CreateDeckRequest {
  name: string;
  description?: string;
  topic?: string;
  difficulty?: "easy" | "medium" | "hard";
  color?: string;
}

export interface UpdateDeckRequest {
  name?: string;
  description?: string;
  topic?: string;
  difficulty?: "easy" | "medium" | "hard";
  color?: string;
}

export interface CreateFlashcardRequest {
  front: string;
  back: string;
  difficulty?: "easy" | "medium" | "hard";
}

export interface CreateBulkFlashcardsRequest {
  flashcards: CreateFlashcardRequest[];
}

export interface StudyCardRequest {
  card_id: string;
  difficulty: "easy" | "medium" | "hard";
  time_spent: number; // in seconds
  correct: boolean;
}

export interface StartStudySessionRequest {
  deck_id: string;
}

export interface EndStudySessionRequest {
  total_time: number; // in seconds
  studied_cards: StudyCardRequest[];
}

export interface RateFlashcardRequest {
  rating: "hard" | "okay" | "easy";
}

export interface RateFlashcardResponse {
  message: string;
  rating: string;
  interval_days: number;
  next_review: string;
}

class FlashcardAPI {
  // Deck operations
  async createDeck(request: CreateDeckRequest): Promise<Deck> {
    const response = await apiService.createDeck(request);
    if (response.error) throw new Error(response.error);
    return response.data!;
  }

  async getDecks(): Promise<Deck[]> {
    const response = await apiService.getDecks();
    if (response.error) throw new Error(response.error);
    return response.data!;
  }

  async getDeck(deckId: string): Promise<Deck> {
    const response = await apiService.getDeck(deckId);
    if (response.error) throw new Error(response.error);
    return response.data!;
  }

  async updateDeck(deckId: string, request: UpdateDeckRequest): Promise<Deck> {
    const response = await apiService.updateDeck(deckId, request);
    if (response.error) throw new Error(response.error);
    return response.data!;
  }

  async deleteDeck(deckId: string): Promise<void> {
    const response = await apiService.deleteDeck(deckId);
    if (response.error) throw new Error(response.error);
  }

  // Flashcard operations
  async createFlashcard(
    deckId: string,
    request: CreateFlashcardRequest
  ): Promise<Flashcard> {
    const response = await apiService.createFlashcard(deckId, request);
    if (response.error) throw new Error(response.error);
    return response.data!;
  }

  async createBulkFlashcards(
    deckId: string,
    request: CreateBulkFlashcardsRequest
  ): Promise<Flashcard[]> {
    const response = await apiService.createBulkFlashcards(deckId, request);
    if (response.error) throw new Error(response.error);
    return response.data!;
  }

  async getFlashcards(deckId: string): Promise<Flashcard[]> {
    const response = await apiService.getFlashcards(deckId);
    if (response.error) throw new Error(response.error);
    return response.data!;
  }

  async getStudyCards(deckId: string, limit?: number): Promise<Flashcard[]> {
    const response = await apiService.getStudyCards(deckId, limit);
    if (response.error) throw new Error(response.error);
    return response.data!;
  }

  async deleteFlashcard(deckId: string, flashcardId: string): Promise<void> {
    // TODO: Add this endpoint to the API service when backend supports it
    console.log(`Would delete flashcard ${flashcardId} from deck ${deckId}`);
    // For now, just simulate the API call
    await new Promise(resolve => setTimeout(resolve, 500));
  }

  async rateFlashcard(
    deckId: string,
    flashcardId: string,
    request: RateFlashcardRequest
  ): Promise<RateFlashcardResponse> {
    const response = await apiService.rateFlashcard(deckId, flashcardId, request);
    if (response.error) throw new Error(response.error);
    return response.data!;
  }

  // Study session operations
  async startStudySession(
    request: StartStudySessionRequest
  ): Promise<StudySession> {
    const response = await apiService.startStudySession(request);
    if (response.error) throw new Error(response.error);
    return response.data!;
  }

  async endStudySession(
    sessionId: string,
    request: EndStudySessionRequest
  ): Promise<StudySession> {
    const response = await apiService.endStudySession(sessionId, request);
    if (response.error) throw new Error(response.error);
    return response.data!;
  }

  // Statistics
  async getFlashcardStats(): Promise<FlashcardStats> {
    const response = await apiService.getFlashcardStats();
    if (response.error) throw new Error(response.error);
    return response.data!;
  }
}

// Create and export a single instance
export const flashcardAPI = new FlashcardAPI();
